src/sys/unix.rs

```diff
@@ -1391,6 +1391,134 @@ pub(crate) const fn to_mreqn(
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct PathMtuDiscoveringModeV4(libc::c_int);
+
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
+impl PathMtuDiscoveringModeV4 {
+    /// Path MTU discovery is not done.
+    pub const DONT: Self = Self(libc::IP_PMTUDISC_DONT);
+    /// Path MTU discovery is done according to route setting.
+    pub const WANT: Self = Self(libc::IP_PMTUDISC_WANT);
+    /// Path MTU discovery
+    pub const DO: Self = Self(libc::IP_PMTUDISC_DO);
+    /// Set DF bit but ignore Path MTDU
+    pub const PROBE: Self = Self(libc::IP_PMTUDISC_PROBE);
+    /// Use interface MTU.
+    /// It ignore destination PMTU and does not set DF flag.
+    /// Incomming ICMP frag_needed notifications on this socket will be ignored in
+    /// order to prevent accepting spoofed ones.
+    pub const INTERFACE: Self = Self(libc::IP_PMTUDISC_INTERFACE);
+    /// Almost like `MtuDiscoveringModeV4::Interface` but authorize fragmented packet
+    /// if they do not saturate the interface MTU.
+    pub const OMIT: Self = Self(libc::IP_PMTUDISC_OMIT);
+}
+
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
+impl From<c_int> for PathMtuDiscoveringModeV4 {
+    fn from(t: c_int) -> PathMtuDiscoveringModeV4 {
+        Self(t)
+    }
+}
+
+#[cfg(target_os = "linux")]
+#[cfg_attr(docsrs, doc(cfg(target_os = "linux")))]
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
+impl From<PathMtuDiscoveringModeV4> for c_int {
+    fn from(t: PathMtuDiscoveringModeV4) -> c_int {
+        t.0
+    }
+}
+
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
+impl_debug!(
+    PathMtuDiscoveringModeV4,
+    libc::IP_PMTUDISC_DONT,
+    libc::IP_PMTUDISC_WANT,
+    libc::IP_PMTUDISC_DO,
+    libc::IP_PMTUDISC_PROBE,
+    libc::IP_PMTUDISC_INTERFACE,
+    libc::IP_PMTUDISC_OMIT
+);
+
+/// Possibles state for Path Maximum Transmission Unit (PMTU) Discovering of packets received on an IPV6 socket.
+///
+/// It set in the IP packet Header the flag DF also known as "do not fragment".
+///
+/// For UDP it's particullary important to note that during Path MTU discovery is done by the OS,
+/// incomming datagram may be dropped.
+///
+/// This type is freely interconvertible with C's `int` type, however, if a raw
+/// value needs to be provided.
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct PathMtuDiscoveringModeV6(libc::c_int);
+
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
+impl PathMtuDiscoveringModeV6 {
```

Comment:
It involve some code duplication alas but it's code that... never change except for adding new mode since API don't change for Linux.